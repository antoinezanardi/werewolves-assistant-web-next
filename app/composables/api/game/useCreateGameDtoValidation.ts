import { storeToRefs } from "pinia";
import { group as groupByField } from "radash";
import type { ComputedRef, Ref } from "vue";

import { MIN_PLAYERS_IN_GAME, MIN_PLAYERS_IN_GROUP } from "~/composables/api/game/constants/game.constants";
import type { CreateGamePlayerWithGroupDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { Role } from "~/composables/api/role/types/role.class";
import { useRolesStore } from "~/stores/role/useRolesStore";

type RoleWithMinInGame = Role & { minInGame: number };

type UseCreateGameDtoValidation = {
  isMinimumPlayersReached: ComputedRef<boolean>;
  areAllRolesSet: ComputedRef<boolean>;
  doesContainOneVillagerSidedRole: ComputedRef<boolean>;
  doesContainOneWerewolfSidedRole: ComputedRef<boolean>;
  areRolesMinimumPlayersReached: ComputedRef<boolean>;
  isTwoSistersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  isThreeBrothersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  areAdditionalCardsSetForThiefIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForActorIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForAdditionalCardsDependantRoles: ComputedRef<boolean>;
  arePlayerGroupsSetForPrejudicedManipulatorIfPresent: ComputedRef<boolean>;
  canCreateGame: ComputedRef<boolean>;
  gameCreationValidationErrors: ComputedRef<string[]>;
  isRolePresentAndMinimumPlayersReached: (role: RoleWithMinInGame) => boolean;
};

function useCreateGameDtoValidation(createGameDto: Ref<CreateGameDto>): UseCreateGameDtoValidation {
  const rolesStore = useRolesStore();
  const { roles } = storeToRefs(rolesStore);

  const { t } = useI18n();

  const isMinimumPlayersReached = computed<boolean>(() => createGameDto.value.players.length >= MIN_PLAYERS_IN_GAME);

  const areAllRolesSet = computed<boolean>(() => createGameDto.value.players.every(player => player.role.name !== undefined));

  const doesContainOneVillagerSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "villagers"));

  const doesContainOneWerewolfSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "werewolves"));

  const areRolesMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const rolesWithMinimumPlayers = roles.value.filter((role): role is RoleWithMinInGame => role.minInGame !== undefined);

    return rolesWithMinimumPlayers.every(role => isRolePresentAndMinimumPlayersReached(role));
  });

  const isTwoSistersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "two-sisters") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const isThreeBrothersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "three-brothers") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const areAdditionalCardsSetForThiefIfPresent = computed<boolean>(() => {
    const isThiefPresent = createGameDto.value.players.some(player => player.role.name === "thief");
    if (!isThiefPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "thief") === true;
  });

  const areAdditionalCardsSetForActorIfPresent = computed<boolean>(() => {
    const isActorPresent = createGameDto.value.players.some(player => player.role.name === "actor");
    if (!isActorPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "actor") === true;
  });

  const areAdditionalCardsSetForAdditionalCardsDependantRoles = computed<boolean>(() => areAdditionalCardsSetForActorIfPresent.value &&
    areAdditionalCardsSetForThiefIfPresent.value);

  const arePlayerGroupsSetForPrejudicedManipulatorIfPresent = computed<boolean>(() => {
    const isPrejudicedManipulatorPresent = createGameDto.value.players.some(player => player.role.name === "prejudiced-manipulator");
    if (!isPrejudicedManipulatorPresent) {
      return true;
    }
    const minimumGroups = 2;
    const playersWithGroup = createGameDto.value.players.filter(player => player.group !== undefined) as CreateGamePlayerWithGroupDto[];
    const groups = groupByField(playersWithGroup, player => player.group) as Record<string, CreateGamePlayerWithGroupDto[]>;
    const areAllGroupsMinimumPlayersReached = Object.values(groups).every(group => group.length >= MIN_PLAYERS_IN_GROUP);

    return Object.keys(groups).length >= minimumGroups && areAllGroupsMinimumPlayersReached;
  });

  const canCreateGame = computed<boolean>(() => isMinimumPlayersReached.value &&
    areAllRolesSet.value &&
    doesContainOneVillagerSidedRole.value &&
    doesContainOneWerewolfSidedRole.value &&
    areRolesMinimumPlayersReached.value &&
    areAdditionalCardsSetForThiefIfPresent.value &&
    areAdditionalCardsSetForActorIfPresent.value &&
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value);

  const gameCreationValidationErrors = computed<string[]>(() => {
    const errors: Record<string, boolean> = {
      [t("composables.useCreateGameDtoValidation.minimumOfPlayersNotReached")]: !isMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.notAllRolesAreSet")]: !areAllRolesSet.value,
      [t("composables.useCreateGameDtoValidation.noVillagerSidedRole")]: !doesContainOneVillagerSidedRole.value,
      [t("composables.useCreateGameDtoValidation.noWerewolfSidedRole")]: !doesContainOneWerewolfSidedRole.value,
      [t("composables.useCreateGameDtoValidation.twoSistersMinimumPlayersNotReached")]: !isTwoSistersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.threeBrothersMinimumPlayersNotReached")]: !isThreeBrothersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.thiefAdditionalCardsNotSet")]: !areAdditionalCardsSetForThiefIfPresent.value,
      [t("composables.useCreateGameDtoValidation.actorAdditionalCardsNotSet")]: !areAdditionalCardsSetForActorIfPresent.value,
      [t("composables.useCreateGameDtoValidation.prejudicedManipulatorGroupsNotSet")]: !arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value,
    };

    return Object.keys(errors).filter((key: keyof typeof errors) => errors[key]);
  });

  function isRolePresentAndMinimumPlayersReached(role: RoleWithMinInGame): boolean {
    const playersWithRole = createGameDto.value.players.filter(player => player.role.name === role.name);
    const playersWithRoleCount = playersWithRole.length;

    return playersWithRoleCount === 0 || playersWithRoleCount >= role.minInGame;
  }
  return {
    isMinimumPlayersReached,
    areAllRolesSet,
    doesContainOneVillagerSidedRole,
    doesContainOneWerewolfSidedRole,
    areRolesMinimumPlayersReached,
    isTwoSistersRolePresentAndMinimumPlayersReached,
    isThreeBrothersRolePresentAndMinimumPlayersReached,
    areAdditionalCardsSetForThiefIfPresent,
    areAdditionalCardsSetForActorIfPresent,
    areAdditionalCardsSetForAdditionalCardsDependantRoles,
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent,
    canCreateGame,
    gameCreationValidationErrors,
    isRolePresentAndMinimumPlayersReached,
  };
}

export { useCreateGameDtoValidation };