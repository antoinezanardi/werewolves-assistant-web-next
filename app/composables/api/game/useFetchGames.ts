import type { CreateGameFeedbackDto } from "~/composables/api/game/dto/create-game-feedback/create-game-feedback.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { Game } from "~/composables/api/game/types/game.class";
import { useWerewolvesAssistantApi } from "~/composables/api/useWerewolvesAssistantApi";

type UseFetchGames = {
  createGame: (createGameDto: CreateGameDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  getGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  cancelGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  makeGamePlay: (gameId: string, makeGamePlayDto: MakeGamePlayDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  createGameFeedback: (gameId: string, createGameFeedbackDto: CreateGameFeedbackDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
};

function useFetchGames(): UseFetchGames {
  const { fetchWerewolvesAssistantApi } = useWerewolvesAssistantApi();

  async function createGame(createGameDto: CreateGameDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games`, {
        method: "POST",
        body: JSON.stringify(createGameDto),
      });
    } catch {
      return null;
    }
  }

  async function getGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "GET" });
    } catch {
      return null;
    }
  }

  async function cancelGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "DELETE" });
    } catch {
      return null;
    }
  }

  async function makeGamePlay(gameId: string, makeGamePlayDto: MakeGamePlayDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}/play`, {
        method: "POST",
        body: JSON.stringify(makeGamePlayDto),
      });
    } catch {
      return null;
    }
  }

  async function createGameFeedback(gameId: string, createGameFeedbackDto: CreateGameFeedbackDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}/feedback`, {
        method: "POST",
        body: JSON.stringify(createGameFeedbackDto),
      });
    } catch {
      return null;
    }
  }
  return {
    createGame,
    getGame,
    cancelGame,
    makeGamePlay,
    createGameFeedback,
  };
}

export { useFetchGames };