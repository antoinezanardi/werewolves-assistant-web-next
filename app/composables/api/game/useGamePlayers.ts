import type { ComputedRef, Ref } from "vue";

import type { Game } from "~/composables/api/game/types/game.class";
import type { Player } from "~/composables/api/game/types/players/player.class";
import type { RoleName } from "~/composables/api/role/types/role.types";

type UseGamePlayers = {
  sheriffInPlayers: ComputedRef<Player | undefined>;
  getPlayersWithCurrentRole: (role: RoleName) => Player[];
};

function useGamePlayers(game: Ref<Game>): UseGamePlayers {
  const sheriffInPlayers = computed<Player | undefined>(() => game.value.players.find(({ attributes }) => attributes.some(({ name }) => name === "sheriff")));

  function getPlayersWithCurrentRole(role: RoleName): Player[] {
    return game.value.players.filter(player => player.role.current === role);
  }
  return {
    getPlayersWithCurrentRole,
    sheriffInPlayers,
  };
}

export { useGamePlayers };