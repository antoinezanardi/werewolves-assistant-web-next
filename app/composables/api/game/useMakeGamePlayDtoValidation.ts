import type { ComputedRef, Ref } from "vue";

import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { Game } from "~/composables/api/game/types/game.class";

type UseMakeGamePlayDtoValidation = {
  isCurrentGamePlayVoteTypeAndValid: ComputedRef<boolean>;
  isCurrentGamePlayTargetTypeAndValid: ComputedRef<boolean>;
  isCurrentGamePlayChooseCardTypeAndValid: ComputedRef<boolean>;
  isCurrentGamePlayChooseSideTypeAndValid: ComputedRef<boolean>;
  canCurrentPlayBeSkipped: ComputedRef<boolean>;
  canMakeGamePlay: ComputedRef<boolean>;
};

function useMakeGamePlayDtoValidation(makeGamePlayDto: Ref<MakeGamePlayDto>, game: Ref<Game>): UseMakeGamePlayDtoValidation {
  const isCurrentGamePlayVoteTypeAndValid = computed<boolean>(() => {
    if (game.value.currentPlay?.type !== "vote" || makeGamePlayDto.value.votes === undefined) {
      return false;
    }
    const voteInteraction = game.value.currentPlay.source.interactions?.[0];
    if (voteInteraction === undefined) {
      return false;
    }
    const { votes } = makeGamePlayDto.value;

    return votes.length >= voteInteraction.boundaries.min;
  });

  const isCurrentGamePlayTargetTypeAndValid = computed<boolean>(() => {
    if (game.value.currentPlay?.type !== "target" || makeGamePlayDto.value.targets === undefined) {
      return false;
    }

    const targetInteraction = game.value.currentPlay.source.interactions?.[0];
    if (targetInteraction === undefined) {
      return false;
    }
    const { targets } = makeGamePlayDto.value;

    return targets.length >= targetInteraction.boundaries.min;
  });

  const isCurrentGamePlayChooseCardTypeAndValid = computed<boolean>(() => game.value.currentPlay?.type === "choose-card" && makeGamePlayDto.value.chosenCardId !== undefined);

  const isCurrentGamePlayChooseSideTypeAndValid = computed<boolean>(() => game.value.currentPlay?.type === "choose-side" && makeGamePlayDto.value.chosenSide !== undefined);

  const canCurrentPlayBeSkipped = computed<boolean>(() => game.value.currentPlay?.canBeSkipped === true);

  const canMakeGamePlay = computed<boolean>(() => canCurrentPlayBeSkipped.value ||
    isCurrentGamePlayVoteTypeAndValid.value ||
    isCurrentGamePlayTargetTypeAndValid.value ||
    isCurrentGamePlayChooseCardTypeAndValid.value ||
    isCurrentGamePlayChooseSideTypeAndValid.value);

  return {
    isCurrentGamePlayVoteTypeAndValid,
    isCurrentGamePlayTargetTypeAndValid,
    isCurrentGamePlayChooseCardTypeAndValid,
    isCurrentGamePlayChooseSideTypeAndValid,
    canCurrentPlayBeSkipped,
    canMakeGamePlay,
  };
}

export { useMakeGamePlayDtoValidation };