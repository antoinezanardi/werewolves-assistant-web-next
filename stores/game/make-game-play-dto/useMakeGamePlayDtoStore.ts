import { defineStore } from "pinia";

import { MakeGamePlayTargetDto } from "~/composables/api/game/dto/make-game-play/make-game-play-target/make-game-play-target.dto";
import { MakeGamePlayVoteDto } from "~/composables/api/game/dto/make-game-play/make-game-play-vote/make-game-play-vote.dto";
import { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { RoleSides } from "~/composables/api/role/enums/role.enums";
import { StoreIds } from "~/stores/enums/store.enum";

const useMakeGamePlayDtoStore = defineStore(StoreIds.MAKE_GAME_PLAY_DTO, () => {
  const makeGamePlayDto = ref<MakeGamePlayDto>(MakeGamePlayDto.create({}));

  function setMakeGamePlayDto(makeGamePlayDtoValue: MakeGamePlayDto): void {
    makeGamePlayDto.value = MakeGamePlayDto.create(makeGamePlayDtoValue);
  }

  function resetMakeGamePlayDto(): void {
    makeGamePlayDto.value = MakeGamePlayDto.create({});
  }

  function addMakeGamePlayTargetDto(target: MakeGamePlayTargetDto): void {
    if (!makeGamePlayDto.value.targets) {
      makeGamePlayDto.value.targets = [];
    }
    makeGamePlayDto.value.targets.push(MakeGamePlayTargetDto.create(target));
  }

  function removeMakeGamePlayTargetDto(targetId: string): void {
    if (!makeGamePlayDto.value.targets) {
      return;
    }
    const targetIndex = makeGamePlayDto.value.targets.findIndex(target => target.playerId === targetId);
    if (targetIndex !== -1) {
      makeGamePlayDto.value.targets = makeGamePlayDto.value.targets.toSpliced(targetIndex, 1);
    }
    if (makeGamePlayDto.value.targets.length === 0) {
      makeGamePlayDto.value.targets = undefined;
    }
  }

  function addMakeGamePlayVoteDto(vote: MakeGamePlayVoteDto): void {
    if (!makeGamePlayDto.value.votes) {
      makeGamePlayDto.value.votes = [];
    }
    makeGamePlayDto.value.votes.push(MakeGamePlayVoteDto.create(vote));
  }

  function removeMakeGamePlayVoteDto(sourceId: string): void {
    if (!makeGamePlayDto.value.votes) {
      return;
    }
    const voteIndex = makeGamePlayDto.value.votes.findIndex(vote => vote.sourceId === sourceId);
    if (voteIndex !== -1) {
      makeGamePlayDto.value.votes = makeGamePlayDto.value.votes.toSpliced(voteIndex, 1);
    }
    if (makeGamePlayDto.value.votes.length === 0) {
      makeGamePlayDto.value.votes = undefined;
    }
  }

  function setDoesJudgeRequestAnotherVote(doesJudgeRequestAnotherVote: boolean | undefined): void {
    makeGamePlayDto.value.doesJudgeRequestAnotherVote = doesJudgeRequestAnotherVote;
  }

  function setChosenCardId(chosenCardId: string | undefined): void {
    makeGamePlayDto.value.chosenCardId = chosenCardId;
  }

  function setChosenSide(chosenSide: RoleSides | undefined): void {
    makeGamePlayDto.value.chosenSide = chosenSide;
  }
  return {
    makeGamePlayDto,
    setMakeGamePlayDto,
    resetMakeGamePlayDto,
    addMakeGamePlayTargetDto,
    removeMakeGamePlayTargetDto,
    addMakeGamePlayVoteDto,
    removeMakeGamePlayVoteDto,
    setDoesJudgeRequestAnotherVote,
    setChosenCardId,
    setChosenSide,
  };
});

export { useMakeGamePlayDtoStore };