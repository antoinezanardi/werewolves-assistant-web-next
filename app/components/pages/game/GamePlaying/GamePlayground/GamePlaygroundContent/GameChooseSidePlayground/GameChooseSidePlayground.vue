<template>
  <div
    id="game-choose-side-playground"
    class="flex flex-col items-center justify-center"
  >
    <div>
      <GlowCapture class="flex">
        <GlowElement>
          <button
            id="choose-villagers-side-button"
            :aria-label="$t('components.GameChooseSidePlayground.villagersSide')"
            class="border-4 border-transparent glow:border-green-500 me-2 p-2 rounded-lg"
            :class="{ '!border-green-500': makeGamePlayDto.chosenSide === 'villagers' }"
            type="button"
            @click.prevent="handleChooseVillagersSideButtonClick"
          >
            <RoleImage
              id="villagers-side-image"
              class="glow:border-green-500 mb-1"
              :class="{ 'border-green-500': makeGamePlayDto.chosenSide === 'villagers' }"
              definition="normal"
              role-name="villager"
              sizes="200"
            />

            <span class="font-semibold">
              {{ $t("components.GameChooseSidePlayground.villagersSide") }}
            </span>
          </button>
        </GlowElement>

        <GlowElement>
          <button
            id="choose-werewolves-side-button"
            :aria-label="$t('components.GameChooseSidePlayground.werewolvesSide')"
            class="border-4 border-transparent glow:border-red-500 ms-2 p-2 rounded-lg"
            :class="{ '!border-red-500': makeGamePlayDto.chosenSide === 'werewolves' }"
            type="button"
            @click="handleChooseWerewolvesSideButtonClick"
          >
            <RoleImage
              id="werewolves-side-image"
              class="glow:border-red-500 mb-1"
              :class="{ 'border-red-500': makeGamePlayDto.chosenSide === 'werewolves' }"
              definition="normal"
              role-name="werewolf"
              sizes="200"
            />

            <span class="font-semibold">
              {{ $t("components.GameChooseSidePlayground.werewolvesSide") }}
            </span>
          </button>
        </GlowElement>
      </glowcapture>
    </div>
  </div>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";

import RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";
import { useMakeGamePlayDtoStore } from "~/stores/game/make-game-play-dto/useMakeGamePlayDtoStore";

const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
const { makeGamePlayDto } = storeToRefs(makeGamePlayDtoStore);
const { setChosenSide } = makeGamePlayDtoStore;

function handleChooseVillagersSideButtonClick(): void {
  setChosenSide("villagers");
}

function handleChooseWerewolvesSideButtonClick(): void {
  setChosenSide("werewolves");
}
</script>