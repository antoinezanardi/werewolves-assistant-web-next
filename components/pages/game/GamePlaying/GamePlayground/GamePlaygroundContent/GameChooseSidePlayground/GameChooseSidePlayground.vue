<template>
  <div
    id="game-choose-side-playground"
    class="flex flex-col items-center justify-center"
  >
    <div>
      <GlowCapture class="flex">
        <GlowElement>
          <button
            id="choose-villagers-side-button"
            class="border-4 border-transparent glow:border-green-500 me-2 p-2 rounded-lg"
            :class="{ '!border-green-500': makeGamePlayDto.chosenSide === RoleSides.VILLAGERS }"
            type="button"
            @click.prevent="handleChooseVillagersSideButtonClick"
          >
            <RoleImage
              id="villagers-side-image"
              class="glow:border-green-500 mb-1"
              :class="{ 'border-green-500': makeGamePlayDto.chosenSide === RoleSides.VILLAGERS }"
              definition="normal"
              :role-name="RoleNames.VILLAGER"
              sizes="200"
            />

            <span class="font-semibold">
              {{ $t("components.GameChooseSidePlayground.villagersSide") }}
            </span>
          </button>
        </GlowElement>

        <GlowElement>
          <button
            id="choose-werewolves-side-button"
            class="border-4 border-transparent glow:border-red-500 ms-2 p-2 rounded-lg"
            :class="{ '!border-red-500': makeGamePlayDto.chosenSide === RoleSides.WEREWOLVES }"
            type="button"
            @click="handleChooseWerewolvesSideButtonClick"
          >
            <RoleImage
              id="werewolves-side-image"
              class="glow:border-red-500 mb-1"
              :class="{ 'border-red-500': makeGamePlayDto.chosenSide === RoleSides.WEREWOLVES }"
              definition="normal"
              :role-name="RoleNames.WEREWOLF"
              sizes="200"
            />

            <span class="font-semibold">
              {{ $t("components.GameChooseSidePlayground.werewolvesSide") }}
            </span>
          </button>
        </GlowElement>
      </glowcapture>
    </div>
  </div>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";

import RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";
import { RoleNames, RoleSides } from "~/composables/api/role/enums/role.enums";
import { useMakeGamePlayDtoStore } from "~/stores/game/make-game-play-dto/useMakeGamePlayDtoStore";

const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
const { makeGamePlayDto } = storeToRefs(makeGamePlayDtoStore);
const { setChosenSide } = makeGamePlayDtoStore;

function handleChooseVillagersSideButtonClick(): void {
  setChosenSide(RoleSides.VILLAGERS);
}

function handleChooseWerewolvesSideButtonClick(): void {
  setChosenSide(RoleSides.WEREWOLVES);
}
</script>