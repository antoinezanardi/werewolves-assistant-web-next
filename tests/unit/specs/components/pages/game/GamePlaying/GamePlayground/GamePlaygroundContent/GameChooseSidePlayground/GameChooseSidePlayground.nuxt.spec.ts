import { createTestingPinia } from "@pinia/testing";
import type { mount } from "@vue/test-utils";

import { doesElementHaveClass } from "@tests/unit/utils/helpers/happy-dom.helpers";
import GameChooseSidePlayground from "~/components/pages/game/GamePlaying/GamePlayground/GamePlaygroundContent/GameChooseSidePlayground/GameChooseSidePlayground.vue";
import type RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";
import { StoreIds } from "~/stores/enums/store.enum";
import { useMakeGamePlayDtoStore } from "~/stores/game/make-game-play-dto/useMakeGamePlayDtoStore";
import { createFakeGame } from "@tests/unit/utils/factories/composables/api/game/game.factory";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";

describe("Game Choose Side Playground Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameChooseSidePlayground>>;
  const testingPinia = { initialState: { [StoreIds.GAME]: { game: createFakeGame() } } };

  async function mountGameChooseSidePlaygroundComponent(): Promise<ReturnType<typeof mount<typeof GameChooseSidePlayground>>> {
    return mountSuspendedComponent(GameChooseSidePlayground, {
      shallow: false,
      global: {
        plugins: [createTestingPinia(testingPinia)],
        stubs: { RoleImage: true },
      },
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameChooseSidePlaygroundComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Choose Villager Side Button", () => {
    it("should select the villager side when clicked.", async() => {
      const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
      const button = wrapper.find<HTMLButtonElement>("#choose-villagers-side-button");
      await button.trigger("click");

      expect(makeGamePlayDtoStore.setChosenSide).toHaveBeenCalledExactlyOnceWith("villagers");
    });

    it("should have emerald border when side is villagers.", async() => {
      const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
      makeGamePlayDtoStore.makeGamePlayDto.chosenSide = "villagers";
      await nextTick();
      const button = wrapper.find<HTMLButtonElement>("#choose-villagers-side-button");

      expect(doesElementHaveClass(button, "!border-emerald-500")).toBeTruthy();
    });

    describe("Villagers Button Image", () => {
      it("should have emerald border when side is villagers.", async() => {
        const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
        makeGamePlayDtoStore.makeGamePlayDto.chosenSide = "villagers";
        await nextTick();
        const button = wrapper.find<HTMLButtonElement>("#choose-villagers-side-button");
        const image = button.findComponent<typeof RoleImage>("#villagers-side-image");

        expect(doesElementHaveClass(image, "!border-emerald-500")).toBeTruthy();
      });
    });

    describe("Villagers Button Text", () => {
      it("should have translated text when rendered.", async() => {
        const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
        makeGamePlayDtoStore.makeGamePlayDto.chosenSide = "villagers";
        await nextTick();
        const button = wrapper.find<HTMLButtonElement>("#choose-villagers-side-button");

        expect(button.text()).toBe("Villagers side");
      });
    });
  });

  describe("Choose Werewolves Side button", () => {
    it("should select the werewolves side when clicked.", async() => {
      const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
      const button = wrapper.find<HTMLButtonElement>("#choose-werewolves-side-button");
      await button.trigger("click");

      expect(makeGamePlayDtoStore.setChosenSide).toHaveBeenCalledExactlyOnceWith("werewolves");
    });

    it("should have red border when side is werewolves.", async() => {
      const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
      makeGamePlayDtoStore.makeGamePlayDto.chosenSide = "werewolves";
      await nextTick();
      const button = wrapper.find<HTMLButtonElement>("#choose-werewolves-side-button");

      expect(doesElementHaveClass(button, "!border-red-500")).toBeTruthy();
    });

    describe("Werewolves Button Image", () => {
      it("should have red border when side is werewolves.", async() => {
        const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
        makeGamePlayDtoStore.makeGamePlayDto.chosenSide = "werewolves";
        await nextTick();
        const button = wrapper.find<HTMLButtonElement>("#choose-werewolves-side-button");
        const image = button.findComponent<typeof RoleImage>("#werewolves-side-image");

        expect(doesElementHaveClass(image, "!border-red-500")).toBeTruthy();
      });
    });

    describe("Werewolves Button Text", () => {
      it("should have translated text when rendered.", async() => {
        const makeGamePlayDtoStore = useMakeGamePlayDtoStore();
        makeGamePlayDtoStore.makeGamePlayDto.chosenSide = "werewolves";
        await nextTick();
        const button = wrapper.find<HTMLButtonElement>("#choose-werewolves-side-button");

        expect(button.text()).toBe("Werewolves side");
      });
    });
  });
});