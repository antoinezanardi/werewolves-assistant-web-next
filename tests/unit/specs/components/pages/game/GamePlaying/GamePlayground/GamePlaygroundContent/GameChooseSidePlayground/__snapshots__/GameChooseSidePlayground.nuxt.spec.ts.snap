// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Choose Side Playground Component > should match snapshot when rendered. 1`] = `
"<div id="game-choose-side-playground" class="flex flex-col items-center justify-center">
  <div class="flex"><button id="choose-villagers-side-button" aria-label="Villagers side" class="border-4 border-transparent glow:border-emerald-500 me-2 p-2 rounded-lg" type="button" data-pd-ripple="true" style="overflow: hidden; position: relative; will-change: transform; transition: 300ms cubic-bezier(.03,.98,.52,.99); transform: perspective(1000px) rotateX(NaNdeg) rotateY(NaNdeg) scale3d(1, 1, 1);">
      <role-image-stub rolename="villager" sizes="200" definition="normal" id="villagers-side-image" class="glow:border-emerald-500 mb-1"></role-image-stub><span class="font-semibold">Villagers side</span><span role="presentation" aria-hidden="true" data-p-ink="true" data-p-ink-active="false" class="p-ink" pd1="" data-pc-name="ripple" data-pc-section="root"></span>
      <div class="js-tilt-glare" style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden;">
        <div class="js-tilt-glare-inner" style="position: absolute; top: 50%; left: 50%; transform: rotate(180deg) translate(-50%, -50%); opacity: 0; width: 0px; height: 0px; transition: opacity 300ms cubic-bezier(.03,.98,.52,.99);"></div>
      </div>
    </button><button id="choose-werewolves-side-button" aria-label="Werewolves side" class="border-4 border-transparent glow:border-red-500 ms-2 p-2 rounded-lg" type="button" data-pd-ripple="true" style="overflow: hidden; position: relative; will-change: transform; transition: 300ms cubic-bezier(.03,.98,.52,.99); transform: perspective(1000px) rotateX(NaNdeg) rotateY(NaNdeg) scale3d(1, 1, 1);">
      <role-image-stub rolename="werewolf" sizes="200" definition="normal" id="werewolves-side-image" class="glow:border-red-500 mb-1"></role-image-stub><span class="font-semibold">Werewolves side</span><span role="presentation" aria-hidden="true" data-p-ink="true" data-p-ink-active="false" class="p-ink" pd2="" data-pc-name="ripple" data-pc-section="root"></span>
      <div class="js-tilt-glare" style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden;">
        <div class="js-tilt-glare-inner" style="position: absolute; top: 50%; left: 50%; transform: rotate(180deg) translate(-50%, -50%); opacity: 0; width: 0px; height: 0px; transition: opacity 300ms cubic-bezier(.03,.98,.52,.99);"></div>
      </div>
    </button></div>
</div>"
`;
